RecoEgamma/EgammaTools/interface/MVAValueMapProducer.h

```diff
@@ -111,7 +127,15 @@ void MVAValueMapProducer<ParticleType>::produce(edm::StreamID,
                                                 const edm::EventSetup& iSetup) const {
   std::vector<float> auxVariables = variableHelper_.getAuxVariables(iEvent);
 
-  auto src = iEvent.getHandle(src_);
+  auto srcHandle = iEvent.getHandle(srcToken_);
+  auto keysForValueMapsHandle =
+      keysForValueMapsToken_.isUninitialized() ? srcHandle : iEvent.getHandle(keysForValueMapsToken_);
+
+  // check if nothing is wrong with the data format of the candidates
+  if (!validated_ && !srcHandle->empty()) {
+    egammaTools::validateEgammaCandidate((*srcHandle)[0]);
```

Comment:
Is it possible that the first Canditate to get validated is not a reco::GsfElectron? If so, the validation wouldn't be applied to the correct collection.

Reply:
The `ParticleType` template argument is either specialized to [`reco::Photon`](https://github.com/cms-sw/cmssw/blob/b3864aef5f76bde11db583070526ac87607aea33/RecoEgamma/PhotonIdentification/plugins/plugins.cc#L8) or [`reco::GsfElectron`](https://github.com/cms-sw/cmssw/blob/5c3f56ee0db7642e67bdab5530e0781a1fbcec80/RecoEgamma/ElectronIdentification/plugins/plugins.cc#L34), not something higher up in the inheritance hierarchy like `reco::Candidate`.

So if any of the candidates passed to the producer is a `reco::GsfElectron`, the first must be one too. If they are photons, the validation function will do nothing thanks to [this check](https://github.com/cms-sw/cmssw/pull/30052/files#diff-68efcb949db29068bd93c757901dceb3R12). I hope I  understood your concerns correctly!

Reply:
My concern is that if the first candidate is a Photon, then the validation function will do nothing but the ```validated_``` flag is set to true anyhow
Perhaps, ```validateEgammaCandidate(...)``` should return a bool saying whether he has actually validated a reco::GsfElectron

Reply:
I thought this is the meaningful behaviour: if it's a photon, the validation is trivial so the validated_ flag can be set to true. What you are saying is basically for `validateEgammaCandidate` to return whether the validation was non-trivial, and then the producer doesn't set the `validated_` flag if it wasn't. I have to admit that I don't see the advantage of that, but maybe I'm missing something?

Reply:
Ok, thank you Jonas. Indeed I misunderstood the behaviour at first (too much time passed since I inspected this code...).

The validation will run for every instantiation of the MVAValueMapProducer separately, any of those instantiations will get validated independently, and the source for those instantiation will remain the same, so no electron and photons having to get validated in the same module. Therefore, it is correct that the validated_ flag is switched to true at the first check, which will be on the only ParticleType to be inspected in that module.